keras_cv/models/object_detection/yolo_v8/yolo_v8_detector.py

```diff
@@ -0,0 +1,538 @@
+        ww_grid, hh_grid = tf.meshgrid(ww_centers, hh_centers)
+        grid = tf.cast(
+            tf.reshape(tf.stack([hh_grid, ww_grid], 2), [-1, 1, 2]),
+            tf.float32,
+        )
+        anchors = tf.expand_dims(base_anchors * [stride, stride], 0) + grid
+        anchors = tf.reshape(anchors, [-1, 2])
+        all_anchors.append(anchors)
+        all_strides.append(tf.repeat(stride, anchors.shape[0]))
+
+    all_anchors = tf.cast(tf.concat(all_anchors, axis=0), tf.float32)
+    all_strides = tf.cast(tf.concat(all_strides, axis=0), tf.float32)
+
+    all_anchors = all_anchors / all_strides[:, None]
+
+    # Swap the x and y coordinates of the anchors.
+    all_anchors = tf.concat(
+        [all_anchors[:, 1, tf.newaxis], all_anchors[:, 0, tf.newaxis]], axis=-1
+    )
+    return all_anchors, all_strides
+
+
+def path_aggregation_fpn(features, depth=3, name=None):
+    # yolov8
+    # 9: p5 1024 ---+----------------------+-> 21: out2 1024
+    #               v [up 1024 -> concat]  ^ [down 512 -> concat]
+    # 6: p4 512 --> 12: p4p5 512 --------> 18: out1 512
+    #               v [up 512 -> concat]   ^ [down 256 -> concat]
+    # 4: p3 256 --> 15: p3p4p5 256 --------+--> 15: out0 128
+    # features: [p3, p4, p5]
+    channel_axis = -1
+    upsamples = [features[-1]]
+    # upsamples: [p5], features[:-1][::-1]: [p4, p3] -> [p5, p4p5, p3p4p5]
+    for id, feature in enumerate(features[:-1][::-1]):
+        size = tf.shape(feature)[1:-1]
+        x = tf.image.resize(upsamples[-1], size, method="nearest")
+        x = tf.concat([x, feature], axis=channel_axis)
+
+        out_channel = feature.shape[channel_axis]
+        x = apply_csp_block(
+            x,
+            channels=out_channel,
+            depth=depth,
+            shortcut=False,
+            activation="swish",
+            name=f"{name}_p{len(features) + 1 - id}",
+        )
+        upsamples.append(x)
+
+    downsamples = [upsamples[-1]]
+    # downsamples: [p3p4p5], upsamples[:-1][::-1]:
+    # [p4p5, p5] -> [p3p4p5, p3p4p5 + p4p5, p3p4p5 + p4p5 + p5]
+    for id, ii in enumerate(upsamples[:-1][::-1]):
+        cur_name = f"{name}_c3n{id + 3}"
+        x = apply_conv_bn(
+            downsamples[-1],
+            downsamples[-1].shape[channel_axis],
+            kernel_size=3,
+            strides=2,
+            activation="swish",
+            name=f"{cur_name}_down",
+        )
+        x = tf.concat([x, ii], axis=channel_axis)
+
+        out_channel = ii.shape[channel_axis]
+        x = apply_csp_block(
+            x,
+            channels=out_channel,
+            depth=depth,
+            shortcut=False,
+            activation="swish",
+            name=cur_name,
+        )
+        downsamples.append(x)
+    return downsamples
+
+
+def yolov8_head(
+    inputs,
+    num_classes,
+    bbox_len=64,
+    name="yolov8_head",
+):
+    outputs = []
+    reg_channels = max(16, bbox_len, inputs[0].shape[-1] // 4)
```

Comment:
This is a bit inscrutable. I guess we don't want more than 16!

Reply:
Updated and added a comment.